src/include/txn_inline.h

```diff
@@ -991,8 +994,22 @@ __wt_txn_read_upd_list_internal(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt,
 
         upd_visible = __wt_txn_upd_visible_type(session, upd);
 
-        if (upd_visible == WT_VISIBLE_TRUE)
+        if (upd_visible == WT_VISIBLE_TRUE) {
+            /*
+             * Special case: We previously found a prepared update, check if the visible update we
+             * found has the same transaction id, if it does it must not be visible as it is part of
+             * the same transaction as the previous prepared updated.
+             *
+             * This indicates that the read, which is configured to ignore prepared updates raced
+             * with the commit of the same prepared transaction. If we were to return this update
+             * then the reader would see a partially committed transaction which is bad.
+             */
+            if (prepare_txnid != WT_TXN_NONE && upd->txnid == prepare_txnid) {
```

Comment:
What about the prepared updates across keys? Do we want to guarantee the reader to see the consistent results?

The situation may be:

We see the update older than prepared update in key 1 and see the just committed prepared update in key 2.

Reply:
We don't guarantee consistent results across keys, see SERVER-48002 for more details.